rfcs/20230517-forward-compatibility-testing.md

```diff
@@ -0,0 +1,99 @@
+# StableHLO Forward Compatibility Testing
+
+## Background
+
+StableHLO needs to verify both forward and backward compatibility. Currently we
+check backward compatibility by versioning and serializing
+`stablehlo_legalize_to_vhlo.mlir`, and need something similar for forward
+compatibility. This document makes the following proposal:
+
+**Proposal:** Add tests to compare `stablehlo_legalize_to_vhlo.0_X_0.mlir`
+serialized at HEAD with existing [`stablehlo_legalize_to_vhlo.0_X_0.mlir.bc`](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir.bc)
+test files.
+
+## Requirements
+
+Forward compatibility testing has the following requirements, to ensure that it
+is both useful and practical:
+
+* **R1.** Tests must detect changes in StableHLO that break forward compatibility.
+* **R2.** Tests must detect LLVM revision bumps that break forward compatibility.
+* **R3.** Tests must scale with releases and execute in a reasonable amount
+of time.
+
+## Proposed Design: Statically Test Forward Compatibility
+
+In this design, forward incompatibilities are detected by serializing
+`stablehlo_legalize_to_vhlo.mlir` at head and comparing it to a known-good file
+serialized serialized at previous releases. This approach operates on the
+assumption that at HEAD we are able to produce a byte-identical serialized
+artifact, excluding custom header info and debug locations.
+
+Detecting forward incompatiblities can be accomplished by adding the following
+`RUN` line to the versioned [`stablehlo_legalize_to_vhlo.0_X_0.mlir`](https://github.com/search?q=repo%3Aopenxla%2Fstablehlo+path%3A**%2Fstablehlo_legalize_to_vhlo.0_*&type=code)
+test files.
+
+```bash
+diff %s.bc <(stablehlo-translate --serialize --target=0.X.0)
+```
+
+This test should be run in CI on all PRs to prevent merging incompatible changes
+in StableHLO (_R1_) and LLVM revision bumps (_R2_).
+
+This will not work as is because `bytecode::producer`, `op::location` and
```

Comment:
The latter two can be fixed, the former too by setting a define I believe.

Reply:
I have a patch that adds the ability to specify producer string, will upstream that as a part of upstream forward compatibility tests, and you're right, we can standardize locations in a test file to take care of the latter two.